domain/src/test/java/org/oppia/domain/classify/rules/numberinput/NumericInputIsWithinToleranceRuleClassifierProviderTest.kt

```diff
@@ -0,0 +1,554 @@
+      "c" to POSITIVE_REAL_VALUE_2_5
+    )
+    val exception = assertThrows(IllegalStateException::class) {
+      inputIsWithinToleranceRuleClassifier
+        .matches(answer = POSITIVE_REAL_VALUE_2_5, inputs = inputs)
+    }
+
+    assertThat(exception)
+      .hasMessageThat()
+      .contains("Expected classifier inputs to contain parameter with name 'tol' but had: [x]")
+  }
+
+  @Test
+  fun testRealAnswer_firstIncorrectInput_throwsException() {
+    val inputs = mapOf(
+      "c" to NEGATIVE_REAL_VALUE_3_5,
+      "tol" to POSITIVE_REAL_VALUE_3_5
+    )
+    val exception = assertThrows(IllegalStateException::class) {
+      inputIsWithinToleranceRuleClassifier
+        .matches(answer = POSITIVE_REAL_VALUE_1_5, inputs = inputs)
+    }
+
+    assertThat(exception)
+      .hasMessageThat()
+      .contains("Expected classifier inputs to contain parameter with name 'x' but had: [c, tol]")
+  }
+
+  @Test
+  fun testRealAnswer_secondIncorrectInput_throwsException() {
+    val inputs = mapOf(
+      "x" to NEGATIVE_REAL_VALUE_3_5,
+      "c" to POSITIVE_REAL_VALUE_1_5
+    )
+    val exception = assertThrows(IllegalStateException::class) {
+      inputIsWithinToleranceRuleClassifier
+        .matches(answer = POSITIVE_REAL_VALUE_1_5, inputs = inputs)
+    }
+
+    assertThat(exception)
+      .hasMessageThat()
+      .contains("Expected classifier inputs to contain parameter with name 'tol' but had: [x, c]")
+  }
+
+  @Test
+  fun testRealAnswer_firstStringInput_throwsException() {
+    val inputs = mapOf(
+      "x" to STRING_VALUE_1,
+      "tol" to NEGATIVE_REAL_VALUE_3_5
+    )
+    val exception = assertThrows(IllegalStateException::class) {
+      inputIsWithinToleranceRuleClassifier
+        .matches(answer = POSITIVE_REAL_VALUE_1_5, inputs = inputs)
+    }
+
+    assertThat(exception)
+      .hasMessageThat()
+      .contains("Expected input value to be of type REAL not NORMALIZED_STRING")
+  }
+
+  @Test
+  fun testIntAnswer_missingFirstInput_throwsException() {
+    val inputs = mapOf(
+      "tol" to POSITIVE_INT_VALUE_3
+    )
+    val exception = assertThrows(IllegalStateException::class) {
+      inputIsWithinToleranceRuleClassifier
+        .matches(answer = POSITIVE_INT_VALUE_2, inputs = inputs)
+    }
+
+    assertThat(exception)
+      .hasMessageThat()
+      .contains("Expected classifier inputs to contain parameter with name 'x' but had: [tol]")
+  }
+
+  @Test
+  fun testIntAnswer_missingSecondInput_throwsException() {
+    val inputs = mapOf(
+      "x" to POSITIVE_INT_VALUE_1
+    )
+    val exception = assertThrows(IllegalStateException::class) {
+      inputIsWithinToleranceRuleClassifier
+        .matches(answer = POSITIVE_INT_VALUE_2, inputs = inputs)
+    }
+
+    assertThat(exception)
+      .hasMessageThat()
+      .contains("Expected classifier inputs to contain parameter with name 'tol' but had: [x]")
+  }
+
+  @Test
+  fun testIntAnswer_extraInput_throwsException() {
+    val inputs = mapOf(
```

Comment:
@MohamedMedhat1998 This is the testcase

Reply:
Ok, thanks.  
Btw you didn't have to push the changes. If you are unsure of the code and you don't want to push it, you can simply add it as a code snippet in a comment like this:  
```
  @Test
  fun testIntAnswer_extraInput_throwsException() {
    val inputs = mapOf(
      "c" to POSITIVE_INT_VALUE_2,
      "x" to POSITIVE_INT_VALUE_3,
      "tol" to POSITIVE_INT_VALUE_1
    )
    val exception = assertThrows(IllegalStateException::class) {
      inputIsWithinToleranceRuleClassifier
        .matches(answer = POSITIVE_INT_VALUE_2, inputs = inputs)
    }

    assertThat(exception)
      .hasMessageThat()
      .contains("Expected classifier inputs to contain parameter with name 'tol' but had: [x]")
  }
```
Try removing any assertion related to the exception and simply use an incorrect map in the normal test cases like `testNegativeRealAnswer_negativeRealInput_positiveRealTolerance_answerWithinTolerance` and observe the output

Reply:
okay, thanks

Reply:
I tried this - 
```
  @Test
  fun testIntAnswer_extraInput_throwsException() {
    val inputs = mapOf(
      "c" to POSITIVE_INT_VALUE_2,
      "x" to POSITIVE_INT_VALUE_3,
      "tol" to POSITIVE_INT_VALUE_1
    )
    val matches =
      inputIsWithinToleranceRuleClassifier
        .matches(answer = POSITIVE_INT_VALUE_2, inputs = inputs)

    assertThat(matches).isFalse()
  }
```

And the result is - 
![Screenshot from 2020-09-30 21-25-32](https://user-images.githubusercontent.com/54636525/94709643-8997ba80-0363-11eb-9c6b-24c9f91d8494.png)

Reply:
```
  private fun <T : Throwable> assertThrows(type: KClass<T>, operation: () -> Unit): T {
    try {
      operation()
      fail("Expected to encounter exception of $type")
    } catch (t: Throwable) {
      if (type.isInstance(t)) {
        return type.cast(t)
      }
      // Unexpected exception; throw it.
      throw t
    }
  }
```
@MohamedMedhat1998 I think we are getting an unexpected exception

Reply:
@prayutsu the output is reasonable because 2 (the answer) is between the range (3-1 : 3+1), so, the assertion `assertThat(matches).isFalse()` failed -expeceted behavior-.  
But the other thing we are trying to observe is the extra `pair` in the `map` ("c" to POSITIVE_INT_VALUE_2) didn't throw any exception and I think it is reasonable too.  
If you go to `GenericRuleClassifier.kt` then go to line 35 `retrieveInputObject` function, you will find that we are only checking whether the required parameters ("x" and "tol" in our case) exist or not, if they don't exist, an exception is thrown, else, the code works as expected (even if we have multiple useless parameters). So, do we need to update the current code to throw an exception if there are useless parameters? Or we can ignore writing test cases for this scenario (having useless parameters)?  
@aggarwalpulkit596 @rt4914 What do you think?

Reply:
This is sort of a general "how do we handle exceptional cases" question. In this case, adding a check for extra input helps make things more predictable, but extra parameters don't actually interfere with the classifier's ability to do its job.

I'm also curious to know what others think about this.

Reply:
@MohamedMedhat1998 I think we can add test to check exceptions.

Reply:
> @MohamedMedhat1998 I think we can add test to check exceptions.

But actually, no exception is thrown in this case. So, whenever we add test cases for the exceptions, they fail.

Reply:
@BenHenning Any suggestions here as per above comment?